### PR TITLE
job priority

### PR DESCRIPTION
**Pull request by golanha**
This change is [Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)

### CODE REVIEW

**Review by NassiHarel — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r5, 20 of 34 files at r6, 1 of 4 files at r7, 1 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 24 of 101 files reviewed, 3 unresolved discussions (waiting on @golanha and @yehiyam)

___
*[core/pipeline-driver-queue/config/main/config.base.js, line 13 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH1ZiE-zKOHxZwonsA:-MrH1ZiE-zKOHxZwonsB:b-g7ly73) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/config/main/config.base.js#L13)):*
> ```JavaScript
> 
> config.rest = {
>     port: formatter.parseInt(process.env.REST_PORT, 7000),
> ```

this is HKube GC port



_Code quote:_
```JavaScript
formatter.parseInt(process.env.REST_PORT, 7000),
```
___
*[core/pipeline-driver-queue/lib/persistency/data-store.js, line 34 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH2nggCD55RJurOErB:-MrH2nggCD55RJurOErC:blb758i) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/lib/persistency/data-store.js#L34)):*
> ```JavaScript
>     }
> 
>     async getPreferredJobs() {
> ```

Where is the part that you store the `preference` in the db?

don't you need to query the `queueRunner.preferredQueue.queue` ?



_Code quote:_
```JavaScript
async getPreferredJobs() {
```
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 38 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH3Fbk-rAG2wap2Yv1:-MrH3Fbk-rAG2wap2Yv2:b63wxtq) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L38)):*
> ```JavaScript
>     }
> 
>     async addPreferredJobs({ jobs, position, query }) {
> ```

Where is the part for pushing to the `queueRunner.preferredQueue.queue` ?

**Review by golanha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 24 of 101 files reviewed, 3 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/pipeline-driver-queue/config/main/config.base.js, line 13 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH1ZiE-zKOHxZwonsA:-MrH8BsJEm0FHTnQvKOM:bagndxc) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/config/main/config.base.js#L13)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

this is HKube GC port
</blockquote></details>

Changed to 7100.
___
*[core/pipeline-driver-queue/lib/persistency/data-store.js, line 34 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH2nggCD55RJurOErB:-MrH8YQjA6btU1WBQa_L:bh280x4) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/lib/persistency/data-store.js#L34)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

Where is the part that you store the `preference` in the db?

don't you need to query the `queueRunner.preferredQueue.queue` ?


</blockquote></details>

you are right its not used, its from a previous implementation.  I removed it.



_Code quote:_
```JavaScript
 async getPreferredJobs(
```
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 38 at r8](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrH3Fbk-rAG2wap2Yv1:-MrH8wE0E02tSQFOwVYQ:bhx45qe) ([raw file](https://github.com/kube-hpc/hkube/blob/db62123d4acd9b1db20d6f8fa9a7852d17aef9b6/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L38)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

Where is the part for pushing to the `queueRunner.preferredQueue.queue` ?
</blockquote></details>

Line 68:
                queueRunner.preferredQueue.queue.splice(index, 0, dequeued[0]);

I don't use the enqueue() function because I don't want the INSERT event to be invoked and rearrange the queue order. 



_Code quote:_
```JavaScript
 async addPreferredJobs
```

**Review by yehiyam — CHANGES_REQUESTED** [review]
> Reviewed 2 of 71 files at r1, 61 of 69 files at r3, 1 of 7 files at r5, 6 of 34 files at r6, 1 of 4 files at r7, 4 of 5 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: all files reviewed, 6 unresolved discussions (waiting on @golanha and @NassiHarel)

___
*[core/openapi-spec/swagger/paths/preferred/preferred.yaml, line 61 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHEpKUAKroUKkf9_dj:-MrHEpKUAKroUKkf9_dk:b-rgnevt) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/openapi-spec/swagger/paths/preferred/preferred.yaml#L61)):*
> ```YAML
>   summary: Remover from preferred
>   description: Remover from preferred
>   requestBody:
> ```

I don't think DELETE with body is something that is standart
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 38 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHGufFD2z66RfzUieT:-MrHGufFD2z66RfzUieU:bdpbose) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L38)):*
> ```JavaScript
>     }
> 
>     async addPreferredJobs({ jobs, position, query }) {
> ```

Why is this method async? I can't find any await
If it is really async, then you risk getting conflicts because finding the index and inserting to the queue is not an atomic operation
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 42 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHG4S30HQbq46R_QEA:-MrHG4S30HQbq46R_QEB:b-o7ctj2) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L42)):*
> ```JavaScript
>         const { tag, pipeline, jobId } = query || {};
>         let index;
>         if (position === 'before') {
> ```

use consts for the position values
Maybe even put these in `@hkube/consts` so it can be used in other services (like api-server)

**Review by golanha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 85 of 101 files reviewed, 6 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/openapi-spec/swagger/paths/preferred/preferred.yaml, line 61 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHEpKUAKroUKkf9_dj:-MrI7UyC7ceMg4kzqShv:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/openapi-spec/swagger/paths/preferred/preferred.yaml#L61)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

I don't think DELETE with body is something that is standart
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 38 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHGufFD2z66RfzUieT:-MrHPBdI8RLJtSo_q1pC:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L38)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Why is this method async? I can't find any await
If it is really async, then you risk getting conflicts because finding the index and inserting to the queue is not an atomic operation
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 42 at r9](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MrHG4S30HQbq46R_QEA:-MrI7W9H47UJLdtqrQpH:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/213461c160384e91fc718c1b1fb03054334e34c0/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L42)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

use consts for the position values
Maybe even put these in `@hkube/consts` so it can be used in other services (like api-server)
</blockquote></details>

Done.

File: core/pipeline-driver-queue/lib/service/preferred-jobs.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
+const queuePosition = require('@hkube/consts').queuePositions;
+const queueRunner = require('../queue-runner');
+const validator = require('../validation');
+const InvalidDataError = require('../errors/InvalidDataError');
+class PreferredJobs {
+    getPreferredJobsList() {
+        return queueRunner.preferredQueue.queue;
+    }
+
+    async deletePreferredJobs(jobIds) {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/lib/service/preferred-jobs.js:10]: no async

File: core/pipeline-driver-queue/lib/service/preferred-jobs.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        return false;
+    }
+
+    addPreferredJobs({ addedJobs }) {
+        validator.preference.validatePreferenceRequest({ addedJobs });
+        const { ids, position, query } = addedJobs;
+        const { tag, pipeline, jobId } = query || {};
+        let index;
+        if (position === queuePosition.BEFORE) {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/lib/service/preferred-jobs.js:44]: This code is a little cumbersome.
Try to split to functions and add some comments

File: core/pipeline-driver-queue/lib/validation/preference.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        );
+        const { position, query } = preferenceRequest.addedJobs;
+        if ((position === 'first' || position === 'last')) {
+            if (query) {
+                throw new InvalidDataError('No query is needed in case of position first or last');
+            }
+        }
+        else {
+            const { tag, jobId, pipeline } = query;
+            if ((tag && jobId) || (tag && pipeline) || (pipeline && jobId)) {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/lib/validation/preference.js:20]: what if we will add two more variables? 
```
if ((tag && jobId) || (tag && pipeline) || (pipeline && jobId)
|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)
|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)
) {
```

**Review by golanha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 80 of 103 files reviewed, 9 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 10 at r14](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgji-r14-10:-MrNtgJ63cXyixfUYDk1:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/293771a17de49eb47eb81b5d35ee3c20da4a3172/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L10)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

no async
</blockquote></details>

Done.



_Code quote:_
```JavaScript
 async deletePreferredJobs
```
___
*[core/pipeline-driver-queue/lib/service/preferred-jobs.js, line 44 at r14](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgji-r14-44:-MrNti6-B0KjKmuD6TUR:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/293771a17de49eb47eb81b5d35ee3c20da4a3172/core/pipeline-driver-queue/lib/service/preferred-jobs.js#L44)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

This code is a little cumbersome.
Try to split to functions and add some comments
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/lib/validation/preference.js, line 20 at r14](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgjm-r14-20:-MrR2iFq6YP8qwdsDg6n:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/293771a17de49eb47eb81b5d35ee3c20da4a3172/core/pipeline-driver-queue/lib/validation/preference.js#L20)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

what if we will add two more variables? 
```
if ((tag && jobId) || (tag && pipeline) || (pipeline && jobId)
|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)
|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)|| (pipeline && jobId)
) {
```
</blockquote></details>

Done.

**Review by NassiHarel — COMMENTED** [review]
> Reviewed 3 of 14 files at r10, 5 of 6 files at r13, 2 of 2 files at r14, 5 of 10 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 94 of 103 files reviewed, 3 unresolved discussions (waiting on @yehiyam)

File: core/pipeline-driver-queue/api/rest-api/routes/preferred.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+const RestServer = require('@hkube/rest-server');
+const preferredService = require('../../../lib/service/preferred-jobs');
+
+const routes = () => {
+    const router = RestServer.router();
+
+    router.get('/', async (req, res) => {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/api/rest-api/routes/preferred.js:7]: this is not async

File: core/pipeline-driver-queue/api/rest-api/routes/preferred.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        const response = preferredService.getPreferredJobsList();
+        res.json(response);
+    });
+    router.post('/', async (req, res) => {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/api/rest-api/routes/preferred.js:11]: this is not async

Comment (NassiHarel) [inline core/pipeline-driver-queue/api/rest-api/routes/preferred.js:11]: This is not a best practice, it is confusing...
The user can set the same jobs ids in `addedJobs` and `removedJobs`...
I think you need to add `router.delete` with a body.

File: core/pipeline-driver-queue/config/main/config.base.js [hunk 6]
```diff
@@ -9,6 +9,18 @@ const storageEncoding = process.env.STORAGE_ENCODING || 'bson';
 config.defaultStorage = process.env.DEFAULT_STORAGE || 's3';
 config.clusterName = process.env.CLUSTER_NAME || 'local';
 
+config.rest = {
+    port: formatter.parseInt(process.env.REST_PORT, 7100),
+    prefix: 'api/v1/driver-queue',
+    poweredBy: 'HKube Pipeline Driver queue',
+    bodySizeLimit: process.env.BODY_SIZE_LIMIT || '50mb'
+};
+config.swagger = {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/config/main/config.base.js:18]: Why you need this?

File: core/pipeline-driver-queue/lib/validation/preference.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        this._validator = validator;
+    }
+
+    validatePreferenceRequest(preferenceRequest) {
+        this._validator.validate(
+            this._validator.definitions.updatePreferredRequest,
+            preferenceRequest
+        );
+        const { position, query } = preferenceRequest.addedJobs;
+        if ((position === 'first' || position === 'last')) {
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/lib/validation/preference.js:13]:
```
if ((position === 'first' || position === 'last') && query) {
```

File: core/pipeline-driver-queue/lib/validation/preference.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        }
+        else {
+            const numberOfFields = Object.keys(query).filter(key => query[key] !== undefined).length;
+            if (numberOfFields > 1) {
+                throw new InvalidDataError('Query must contain only one of jobId ,tag ,pipelineName');
+            }
+            const { tag, jobId, pipeline } = query;
+            if (!(tag || pipeline || jobId)) {
+                throw new InvalidDataError('Query must contain one of jobId ,tag ,pipelineName');
+            }
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/lib/validation/preference.js:26]: You can say
```
 if (numberOfFields === 0) {
      throw new InvalidDataError('Query must contain one of jobId ,tag ,pipelineName');
}
```

File: core/pipeline-driver-queue/package.json [hunk 9]
```diff
@@ -29,7 +29,11 @@
     "lodash.isequal": "^4.5.0",
     "lodash.orderby": "^4.6.0",
     "lodash.remove": "^4.7.0",
-    "performance-now": "^2.1.0"
+    "performance-now": "^2.1.0",
```
Comment (NassiHarel) [inline core/pipeline-driver-queue/package.json:32]: Do you need `performance-now`?

**Review by golanha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466)* status: 87 of 104 files reviewed, 9 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/pipeline-driver-queue/package.json, line 32 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-Mq9fiWTfzpWS4ri48fP-r17-32:-MrXR2Nc--gtFciVor1L:b-kkp36a) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/package.json#L32)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

Do you need `performance-now`?
</blockquote></details>

I didn't add it. It is currently  used in scoreing.js
___
*[core/pipeline-driver-queue/config/main/config.base.js, line 18 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgj_-r17-18:-MrXS3mqDnjiGfC4ushP:b-i5ci1x) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/config/main/config.base.js#L18)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

Why you need this?
</blockquote></details>

You are right, I didn't use it.
___
*[core/pipeline-driver-queue/lib/validation/preference.js, line 13 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgjm-r17-13:-MrXTEY1Dc3VYbnPFhaj:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/lib/validation/preference.js#L13)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

```
if ((position === 'first' || position === 'last') && query) {
```
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/lib/validation/preference.js, line 26 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgjm-r17-26:-MrXSOVG-mqTz3GfWjDV:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/lib/validation/preference.js#L26)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

You can say
```
 if (numberOfFields === 0) {
      throw new InvalidDataError('Query must contain one of jobId ,tag ,pipelineName');
}
```
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/api/rest-api/routes/preferred.js, line 7 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgjX-r17-7:-MrXRvoy6o-yATQc802X:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/api/rest-api/routes/preferred.js#L7)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

this is not async
</blockquote></details>

Done.
___
*[core/pipeline-driver-queue/api/rest-api/routes/preferred.js, line 11 at r17](https://reviewable.io/reviews/kube-hpc/hkube/1466#-MqyN-goT6PYFXfGJgjX-r17-11:-Mra6q1d0anIzTMxFMA-:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/9c22ec915526dc3bba2e2521c4456138a5c0ac2d/core/pipeline-driver-queue/api/rest-api/routes/preferred.js#L11)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

This is not a best practice, it is confusing...
The user can set the same jobs ids in `addedJobs` and `removedJobs`...
I think you need to add `router.delete` with a body.
</blockquote></details>

Done.

**Review by NassiHarel — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1466#-:-MrbSwlhC99rm7kandfr:bnfp4nl)* status: 87 of 104 files reviewed, 3 unresolved discussions (waiting on @NassiHarel and @yehiyam)

**Pull request merged**